python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+    trans : bool
+            Should the `jones_transform` be applied?
+            Note: The parameters from a `fit()` model are already transformed.
+    handle : cumlHandle (optional)
+             The cumlHandle to be used.
+    """
+
+    cdef vector[double] vec_loglike
+
+    vec_loglike = _batched_loglike(num_batches, nobs, order, y, x, trans, handle)
+
+    loglike = np.zeros(num_batches)
+    for i in range(num_batches):
+        loglike[i] = vec_loglike[i]
+
+    return loglike
+
+def ll_gf(num_batches, nobs, num_parameters, order, y, x, h=1e-8, trans=True, handle=None):
+    """Computes gradient (via finite differencing) of the batched loglikelihood.
+
+    Parameters:
+    ----------
+    num_batches : int
+                  Number of series
+    nobs : int
+           Number of samples in each series (identical across series)
+    num_parameters : int
+                     The number of parameters per series (p + d + q)
+    order : Tuple[int, int, int]
+            ARIMA Order (p, d, q)
+    y     : array like, shape = (n_samples, n_series)
+            Time series data
+    x     : array
+            dense parameter array, grouped by series, and again by [(mu, ar, ma), (mu, ar, ma)]
+    h     : float
+            The finite-difference stepsize
+    trans : bool
+            Should the `jones_transform` be applied?
+            Note: The parameters from a `fit()` model are already transformed.
+    handle : cumlHandle (optional)
+             The cumlHandle to be used.
+    """
+    pynvtx_range_push("ll_gf")
+    
+    fd = np.zeros(num_parameters)
+
+    grad = np.zeros(len(x))
+
+    # 1st order FD saves 20% runtime.
+    # ll_b0 = ll_f(num_batches, num_parameters, order, y, x, trans=trans)
+    assert (len(x) / num_parameters) == float(num_batches)
+    for i in range(num_parameters):
+        fd[i] = h
+
+        # duplicate the perturbation across batches (they are independent)
+        fdph = np.tile(fd, num_batches)
+
+        # reset perturbation
+        fd[i] = 0.0
+
+        ll_b_ph = ll_f(num_batches, nobs, order, y, x+fdph, trans=trans, handle=handle)
+        ll_b_mh = ll_f(num_batches, nobs, order, y, x-fdph, trans=trans, handle=handle)
+        
+        np.seterr(all='raise')
+        # first derivative second order accuracy
+        grad_i_b = (ll_b_ph - ll_b_mh)/(2*h)
+        # first derivative first order accuracy
+        # grad_i_b = (ll_b_ph - ll_b0)/(h)
+
+        if num_batches == 1:
+            grad[i] = grad_i_b
+        else:
+            assert len(grad[i::num_parameters]) == len(grad_i_b)
+            # Distribute the result to all batches
+            grad[i::num_parameters] = grad_i_b
+
+    pynvtx_range_pop()
+    return grad
+
+def fit(y,
+        order: Tuple[int, int, int],
+        mu0: np.ndarray,
+        ar_params0: List[np.ndarray],
+        ma_params0: List[np.ndarray],
+        opt_disp=-1, h=1e-9):
+    """
```

Comment:
I'm a little bit confused. Shouldn't this `fit` be a method inside the arima class?

Reply:
I have a preference that `fit` should be a free method, which generates a model that is ready for use (i.e., `predict()` and `forecast()`). I'm open to changing it however if someone else feels strongly for the other case (class method).